### PR TITLE
pretrain architectures

### PR DESCRIPTION
**Pull request by svlandeg**
## Description
Adding the pretrain architectures to the registry instead of hard-coding them, so you can do this:

```
[pretraining.objective]
@architectures = "spacy.PretrainVectors.v1"
maxout_pieces = 3
hidden_size = 300
loss = "cosine"
```
or
```
[pretraining.objective]
@architectures = "spacy.PretrainCharacters.v1"
maxout_pieces = 3
hidden_size = 300
n_characters = 4
```
The loss function is stored in `model.attrs["loss"]`.

Not sure about the name & whether they should be in the `architectures` registry, because these two new functions are of type `Callable[["Vocab", "Model"], "Model"]` instead of just `Model`, as they need access to the `vocab` and `tok2vec` objects from the general pipeline.

### Types of change
enhancement

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by honnibal** [comment]
Well generally things in `@architectures` return `Model` instances, and they can take arbitrary arguments. So I think they look okay in that registry to me. Nice to have this task checked off!
